### PR TITLE
Pubdev 4965: Support netezza jdbc driver

### PR DESCRIPTION
**Pull request by Pscheidl**
[PUBDEV JIRA TICKET](https://0xdata.atlassian.net/browse/PUBDEV-4965)

### CODE REVIEW

**Comment by Pscheidl** [comment]
Following scenarios were tested manually during development:

    Absence of Netezza driver should cause a RuntimeException to be thrown. Current expected message is: Connection to Netezza database is not possible due to missing JDBC driver.
    Netezza JDBC level 3 driver is registered with Class.forname when netezza database type is detected from JDBC connection string.

## Absence of Netezza JDBC driver

The exception thrown is an expected outcome. 

```plain
connection_url <- "jdbc:netezza://192.168.56.101:5480/TEST"
> table <- "_T_CHARSET"
> username <- "admin"
> password <- "password"
> data <- h2o.import_sql_table(connection_url, table, username, password)

ERROR: Unexpected HTTP Status code: 500 Server Error (url = http://localhost:54321/99/ImportSQLTable)

java.lang.RuntimeException
 [1] "java.lang.RuntimeException: Connection to Netezza database is not possible due to missing JDBC driver."      
 [2] "    water.jdbc.SQLManager.initializeDatabaseDriver(SQLManager.java:255)"                                     
 [3] "    water.jdbc.SQLManager.importSqlTable(SQLManager.java:51)"                                                
 [4] "    water.api.ImportSQLTableHandler.importSQLTable(ImportSQLTableHandler.java:17)"                           
 [5] "    sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)"                                             
 [6] "    sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)"                           
 [7] "    sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)"                   
 [8] "    java.lang.reflect.Method.invoke(Method.java:498)"                                                        
 [9] "    water.api.Handler.handle(Handler.java:63)"                                                               
[10] "    water.api.RequestServer.serve(RequestServer.java:451)"                                                   
[11] "    water.api.RequestServer.doGeneric(RequestServer.java:296)"                                               
[12] "    water.api.RequestServer.doPost(RequestServer.java:222)"                                                  
[13] "    javax.servlet.http.HttpServlet.service(HttpServlet.java:755)"                                            
[14] "    javax.servlet.http.HttpServlet.service(HttpServlet.java:848)"                                            
[15] "    org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:684)"                                  
[16] "    org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:503)"                              
[17] "    org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1086)"                      
[18] "    org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:429)"                               
[19] "    org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1020)"                       
[20] "    org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:135)"                           
[21] "    org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:154)"                   
[22] "    org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:116)"                         
[23] "    water.JettyHTTPD$LoginHandler.handle(JettyHTTPD.java:192)"                                               
[24] "    org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:154)"                   
[25] "    org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:116)"                         
[26] "    org.eclipse.jetty.server.Server.handle(Server.java:370)"                                                 
[27] "    org.eclipse.jetty.server.AbstractHttpConnection.handleRequest(AbstractHttpConnection.java:494)"          
[28] "    org.eclipse.jetty.server.BlockingHttpConnection.handleRequest(BlockingHttpConnection.java:53)"           
[29] "    org.eclipse.jetty.server.AbstractHttpConnection.content(AbstractHttpConnection.java:982)"                
[30] "    org.eclipse.jetty.server.AbstractHttpConnection$RequestHandler.content(AbstractHttpConnection.java:1043)"
[31] "    org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:865)"                                        
[32] "    org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:240)"                                   
[33] "    org.eclipse.jetty.server.BlockingHttpConnection.handle(BlockingHttpConnection.java:72)"                  
[34] "    org.eclipse.jetty.server.bio.SocketConnector$ConnectorEndPoint.run(SocketConnector.java:264)"            
[35] "    org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:608)"                        
[36] "    org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:543)"                         
[37] "    java.lang.Thread.run(Thread.java:745)"                                                                   

Error in .h2o.doSafeREST(h2oRestApiVersion = h2oRestApiVersion, urlSuffix = page,  : 
  

ERROR MESSAGE:

Connection to Netezza database is not possible due to missing JDBC driver.
```

**Comment by Pscheidl** [comment]
## Re-testing

1. Download [Netezza Emulator](https://www.ibm.com/developerworks/community/wikis/home?lang=en_%29%3A#!/wiki/W4c715f465f89_464f_9fd9_b1d36a3c54f8/page/Requirements%20for%20the%20IBM%20Netezza%20Software%20Emulator%20-%20for%20Developers). 
2. Set-up virtual machine with Netezza emulator. Both Oracle VirtualBox and VMWare work. 
3. Configure network settings of the virtual machine to be able to connect from host environment.
4. Run h2o-3 and the virtual machine
5. Execute h2o.import_sql_table(...)

### Downloading Netezza Emulator
An IBM account is required. Registration is for free. First, a download manager is downloaded. Then, by means of the download manager, Netezza Emulator VM is downloaded. Estimated size is 4 Gb.

To run the downloaded virtual machine, Oracle VirtualBox has been proved to work. There is also [VMWare tutoria](http://dwgeek.com/install-vmware-player-netezza-emulator.html/)l available.

Import the virtual machine image into VirtualBox by clicking on File -> Import appliance.

### Configuring virtual machine network (VirtualBox)
In order to be able to make requests from the host machine to the virtual machine, a host-only network must be configured. After running Oracle VirtualBox, go to File -> Preferences -> Network. After you find yourself in the network submenu, click on "Host-only network" tab. Click on green icon with a plus sign on the right to add "vboxnet0" network. Click OK.

After "vboxnet0" host-only network is set-up, go to preferences of an imported virtual machine with Netezza emulator. Enter the "network" submenu and make sure there is exactly one network interface controller enabled. Set it to "Host-only network" and select network with name "vboxnet0". Click OK.

### Running h2o-3 and the virtual machine

Running the virtual machine is sufficient. No need to make any changes. In case of some older virtual machines, IPTables have to be stopped. If this is the case, log in as root (root/netezza) and turn off iptables by issuing `/etc/init.d/iptables stop` command.

h2o-3 must be ran with Netezza JDBC driver linked on classpath. Adding JDBC driver on classpath during the startup process of h2o.jar is described in [H2O docs]. 

The JDBC driver can be found in the virtual machine in `/nz/kit{kit-version}/sbin/nzjdbc3.jar` . However, it has been extracted and attached to [https://0xdata.atlassian.net/browse/PUBDEV-4965](PUBDEV-4965) ticket.

### Execute h2o.import_sql_table

Test Netezza JDBC connection. E.g. from R:

> connection_url <- "jdbc:netezza://192.168.56.101:5480/TEST"
> table <- "_T_CHARSET"
> username <- "admin"
> password <- "password"
> data <- h2o.import_sql_table(connection_url, table, username, password)

The IP address in JDBC connection URL is the default address assigned in "vboxnet0" network by VirtualBox. Issue `ifconfig` in the VirtualBox to find out your specific address.

`TEST` is one of the pre-defined schemas in Netezza emulator. User credentials are default to the Netezza emulator aswell.

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 1]
```diff
@@ -73,6 +79,7 @@ SELECT ROW_NUMBER() OVER () AS RowNum_, <table>.* FROM <table>
         rs = stmt.executeQuery("SELECT COUNT(1) FROM " + table);
         rs.next();
         numRow = rs.getLong(1);
+        rs.close();
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: good catch!

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 2]
```diff
@@ -13,11 +13,16 @@
 import static water.fvec.Vec.makeCon;
 
 public class SQLManager {
-  
-  final static String TEMP_TABLE_NAME = "table_for_h2o_import";
+
+  private final static String TEMP_TABLE_NAME = "table_for_h2o_import";
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: General note (In this case it is okay): we need to be careful about changing the access modifiers H2O is oftentimes embedded in 3rd party product and making something private can cause issues for the users.

Comment (Pscheidl) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: Thank you. Noted :)

File: h2o-core/src/main/java/water/jdbc/SQLManager.java (hunk 1 above)
Comment (Pscheidl) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: In fact, this one is required for Netezza driver to work, as some drivers do not support more than one connection at a time. We could do a conn.getMetadata()... check, but in this case such check is completely in vain and the resultSet can be closed right away.

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 3]
```diff
@@ -197,6 +204,22 @@ public void compute2() {
...
   }
 
+  /**
+   * Initializes database driver for databases with JDBC driver version lower than 4.0
+   *
+   * @param databaseType Name of target database from JDBC connection string
+   */
+  private static void initializeDatabaseDriver(String databaseType) {
+    switch (databaseType) {
+      case NETEZZA_DB_TYPE:
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: Maybe this check should be case-insensitive ?

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 4]
```diff
@@ -43,11 +48,12 @@
...
         SELECT ROW_NUMBER() OVER () AS RowNum_, <table>.* FROM <table>
      ) QUALIFY RowNum_ BETWEEN x and x+y;
-    */
-    String db_sys = connection_url.split(":",3)[1];
-    final boolean needFetchClause = db_sys.equals("oracle") || db_sys.equals("sqlserver");
-    int catcols = 0, intcols = 0, bincols = 0, realcols = 0, timecols = 0, stringcols = 0; 
-    final int numCol; 
+     */
+    String databaseType = connection_url.split(":", 3)[1];
+    initializeDatabaseDriver(databaseType);
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: These checks are originally case-sensitive. Maybe db_sys should be converted to lower case to make any ongoing matching case-insensitive.

**Pull request merged**